### ISSUE TITLE
Styling issues in autocomplete widget

### ISSUE THREAD

**Issue by nategranatir** [sehmaschine/django-grappelli#764]
Hello,

I'm developing an app using Django 1.9.7 and Grappelli 2.8.1 and noticed some strangeness in the display of the autocomplete widget. It's the first time I've used the autocomplete widget with Grappelli, so not sure what may be causing this.

First, the display value of the widget is shown not only inside the textbox but also within/behind it; see screenshot to see what I mean.
![image](https://cloud.githubusercontent.com/assets/3166184/16969444/8bf54b14-4dda-11e6-916a-935f415bd5d4.png)

Second the textbox is two lines instead of one, with the current value in the first line and the second line where the cursor goes. This could be the expected behavior, I'm not certain, but it looks funny:
![image](https://cloud.githubusercontent.com/assets/3166184/16969499/e1f037ea-4dda-11e6-863f-ff2487dba967.png)

Here's the HTML behind the first widget displayed in the screenshot above, if this helps shine light onto what might be going wrong.
```html
<div class="grp-autocomplete-wrapper-m2m">
	<ul class="grp-repr">
		<li class="grp-repr">
			<a class="grp-m2m-remove" href="javascript://">Network - Syndicated</a>
		</li>
		<li class="grp-search">
			<input id="id_TVShow_tags-0-taghierarchy-autocomplete" type="text" class="vTextField ui-autocomplete-input" value="" autocomplete="off">
		</li>
	</ul>

    <input class="vForeignKeyRawIdAdminField grp-autocomplete-hidden-field" id="id_TVShow_tags-0-taghierarchy" name="TVShow_tags-0-taghierarchy" type="text" value="97" tabindex="-1" readonly="readonly">
    <a href="/admin/syndicated_dictionaries/taghierarchy/?_to_field=id" class="related-lookup" id="lookup_id_TVShow_tags-0-taghierarchy" title="Lookup"></a>
    <a class="grp-related-remove" id="remove_id_TVShow_tags-0-taghierarchy" href="javascript://" onclick="return removeRelatedObject(this);" style="display: inline;"></a>
    <div class="grp-loader" style="display: none;">loader</div>&nbsp;<strong>Network - Syndicated</strong>
</div>
```

Thanks.

**Comment by sehmaschine** [comment]
Not related to the styling issue, but ... you are using an M2M-autocomplete with inlines, right? Is that really necessary?

**Comment by nategranatir** [comment]
I am... 😬  Is there a simpler way to accomplish this that I missed? We have a M2M relationship to a table that has a couple hundred items, for which autocomplete seems to be the easiest way to add those relationships, much better than the default listbox or dropdown list.

**Comment by sehmaschine** [comment]
M2M seems good – alternatively, you could use a FK with an intermediate table (and inlines with the admin interface). The 2nd approach is better, if you need additional fields (e.g. position). But you mix both options ...

**Comment by nategranatir** [comment]
We do have other attributes on the tags. I guess I don't understand how using FKs with an intermediate table would be any different, functionally, than M2M. I thought this was the same except that Django manages the intermediate table for you if you use M2M.

Is there something about this structure that may be causing the weird styling issues?

**Comment by parsch** [comment]
The styling issue 1 comes from `<div class="grp-loader" style="display: none;">loader</div>&nbsp;<strong>Network - Syndicated</strong>`, there should definitely be no `<strong>`. I have no idea where this might come from. Can you provide the HTML of the second example, too?

**Comment by sehmaschine** [comment]
About the database setup: FK with an intermediate table means that you have inlines with a foreign key ... but you have inlines with M2M (that's done twice). If you post your models here I can tell you more about this.

**Comment by sehmaschine** [comment]
Concerning the styling issue: Is it possible that either your `related_label` and/or your `__str__` method contains HTML characters? If so, that causes the error.

**Comment by nategranatir** [comment]
Sorry for the delay in responding. I've done some investigating and hopefully it will shed some light on the issues (I personally am no closer to understanding what's wrong.)

1) @parsch I was using the same instance for both screenshots, so the HTML should be the same in both cases; unless you were referring to the HTML of a different portion of the page?

2) @sehmaschine I don't have HTML characters in either related_label, __str__ or __unicode__; it's just outputting fields that should contain only unicode characters. 

3) Because you mentioned related_label, I added one for our tags and discovered that the string representation inside the box will use related_label, but the weird extra string representation hiding behind the box is using __str__ (previously I had no related_label defined, so __str__ was used for both.) I'm not sure if this suggests what might be wrong in causing the extra text to display.

4) I played around with the related_label text to try and figure out why the display is weird, expanding to two lines. It seems like it *may* have something to do with the length of the text displayed. After playing around with it for a bit, I've taken screenshots of two examples. The only difference between the two is that I made the text of the related_label in the second one slightly longer, and in that case we get the 2-line textbox, in the shorter-text example I get it how (I assume) it's intended to be displayed. I got the same results in both cases in Chrome, Firefox and Safari on my Mac.

![autocomplete_small](https://cloud.githubusercontent.com/assets/3166184/17221990/be902502-54bb-11e6-8130-ebe757e604e6.jpg)

![autocomplete_large](https://cloud.githubusercontent.com/assets/3166184/17221994/c48b7c9a-54bb-11e6-8847-8e2dbd7c689e.jpg)

**Comment by sehmaschine** [comment]
3) that´s the intended behaviour.
4) the multiline textbox is fine for M2M autocompletes.

however, I still think you should redefine the relationship with your database using foreign keys instead of many-to-many relations. could you please send the significant part of your models as well as your admin definitions?

your models should looke something like this ...

```
class Parent(models.Model):
    (fields)

class Tag(models.Model):
    name = models.CharField(u"Name", max_length=100)
    slug = models.SlugField(u"Slug", max_length=100, db_index=True, blank=True)
    (additional fields)

class TagToParent(models.Model):
    parent = models.ForeignKey(Parent, verbose_name=u"Parent", related_name="USE_A_GOOD_NAME")
    tag = models.ForeignKey(Tag, verbose_name=u"Tag", related_name="USE_A_GOOD_NAME")
    (additional fields)
```

and your admin file ...

```
class TagInline(admin.TabularInline):
    model = TagToParent
    raw_id_fields = ("tag",)
    autocomplete_lookup_fields = {
        "fk": ["tag"],
    }

class ParentOptions(admin.ModelAdmin):
    inlines = [TagInline]
```

**Comment by nategranatir** [comment]
Yeah, that looks very similar to what we have, except that Django creates the intermediate table, TVShow.tags.through. Here are the relevant sections of our models (removed a bunch of irrelevant stuff):

```
class TagHierarchy(models.Model):
    dict = models.ForeignKey(Dictionary)
    tag = models.ForeignKey(Tag)
    parent_tag = models.ForeignKey(Tag, related_name='is_parent_tag', null=True, blank=True)

    @staticmethod
    def autocomplete_search_fields():
        return ("tag__tag_name__icontains", "tag__group__group_name__icontains")

    def related_label(self):
        return self.tag.kairos_name()

class InterestDictionary(models.Model):
    interest_name = models.CharField(max_length=500, unique=True, validators=[no_double_quotes])
    interest_id = models.CharField(max_length=20, null=True, blank=True)
    tags = models.ManyToManyField(TagHierarchy, blank=True)
```

```
class TagHierarchyInline(admin.TabularInline):
    form = TagHierarchyInlineForm
    model = TVShow.tags.through
    extra = 0
    verbose_name = 'Add/Remove Tag'
    verbose_name_plural = 'Add/Remove Tags'
    classes = ['grp-collapse grp-closed']

    # For grappelli autocomplete lookup feature
    raw_id_fields = ('taghierarchy',)
    autocomplete_lookup_fields = {
        'm2m': ['taghierarchy'],
    }

class TVShowAdmin(admin.ModelAdmin):
    form = TVShowForm
    list_display = ('interest_name', 'release_date', 'end_date', 'retired', 'searchable_titles')
    fieldsets = [
        ...blah...
        ('Add/Remove Tags', {'classes': ('placeholder TVShow_tags-group',), 'fields': ()}),
        ...blah...
    ]
    inlines = [TagHierarchyInline, TwitterHandleInline]
```

**Comment by sehmaschine** [comment]
No, it does NOT look similar. As mentioned before, you are combining an inline with a many-to-many relation.  I'm not exactly sure what you're trying to achieve here ... and it is definitely not true that "Django creates the intermediate table" (it only does so if you explicitely tell Django).

I'm closing this ticket for now since I'm pretty sure the issue is related to your setup. I would try to clean the models first. And if the issue still exists with a proper database setup, feel free to reopen this ticket.

**Comment by nategranatir** [comment]
I'm not trying to be difficult here, I genuinely don't understand how what you're describing is functionally any different from what I did. Django created the table below automatically, because I defined the ManyToManyField in TVShow (TVShow inherits from InterestDictionary). 
![image](https://cloud.githubusercontent.com/assets/3166184/17332502/97b870d8-5894-11e6-8a78-b1a6e5b59dde.png)


It's a table that has three fields: an id, a foreign key to the TVShow model, and a foreign key to the TagHierarchy model - just like in your example of the TagToParent intermediate table. My inline is based on that model, the automatically created one, accessed via TVShow.tags.through, just like your inline is based on TagToParent.

What I really don't understand about your suggestions are:
- If my use of ManyToManyField is inappropriate here, in what situation _would_ you use ManyToManyField? A Tag can be applied to many Interests, and an Interest can have many tags. I don't see why I wouldn't use ManyToManyField.
- Given that I'm using a ManyToManyField, how _should_ I display that in the admin site? The default display is a select-box, which is cumbersome to use for a large list of Tags. I figured it would be easier, then, to use an inline so tags could be individually added and removed; is there a way to do this without an inline?

I will try to create some test models the way you described and see if I get any difference in behavior.

**Comment by sehmaschine** [comment]
an M2M is essentially the same as an intermediary table with a foreign key (with the 2nd offering more possibilities). so, admin-wise, you either go with a SINGLE M2M autocomplete field ... OR ... with inlines and foreign keys.

many-to-many is fine when using tags. but again, you're not using M2M ... your doing an M2M-M2M.

**Comment by nategranatir** [comment]
Aha! We finally discovered the problem, and understand a little better what you were trying to tell us. The issue was with this line in the intermediate table inline:
```
autocomplete_lookup_fields = {
        'm2m': ['taghierarchy'],
    }
```
It should be:
```
autocomplete_lookup_fields = {
        'fk': ['taghierarchy'],
    }
```
Once I made that change, the display works as intended! Everything seems to be in order - both the two-line display issue and the multiple-string-display issue are resolved. I appreciate your help. It was not clear to me until now that this was the line causing the problems, generating the additional m2m relationship; I misunderstood this line, thinking it was intended to describe how the two tables were related.

Hopefully this thread will be useful to anyone else who makes this mistake.
